### PR TITLE
KAFKA-18235: can't build release tar due to broken genConnectOpenAPIDocs task

### PR DESCRIPTION
**Pull request by m1a2st**
Jira: https://issues.apache.org/jira/browse/KAFKA-18235

add `javax.ws.rs:javax.ws.rs-api:2.1.1` for swagger 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by m1a2st** [comment]
It can pass on my local 
CleanShot 2024-12-14 at 07 28 26@2x

File: build.gradle [hunk 1]
```diff
@@ -3533,6 +3533,7 @@ project(':connect:runtime') {
     implementation libs.classgraph
     implementation libs.mavenArtifact
     implementation libs.swaggerAnnotations
+    implementation libs.javaxWsRs
```
Comment (chia7712) [inline build.gradle]: could you please take a look at my suggested approach in the jira? we should NOT add it back to production scope. it is used by swagger scope only

**Comment by cshannon** [comment]
As I notated on the issue in Jira, it looks like we should be able to configure swagger for Jakarta:

https://github.com/swagger-api/swagger-core/blob/master/modules/swagger-gradle-plugin/README.md#jakarta-namespace-support
https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Getting-started#note-jakarta-namespace-support-since-version-217

**Comment by m1a2st** [comment]
I use `./gradlew clean releaseTarGz` in my local,
image

**Comment by chia7712** [comment]
@m1a2st Could you please test the approach suggested by @cshannon?

File: build.gradle [hunk 2]
```diff
@@ -3540,6 +3540,7 @@ project(':connect:runtime') {
     // We use this library to generate OpenAPI docs for the REST API, but we don't want or need it at compile
     // or run time. So, we add it to a separate configuration, which we use later on during docs generation
     swagger libs.swaggerJaxrs2
```
Comment (chia7712) [inline build.gradle]: @cshannon you are right - it is better to use `swagger-jaxrs2-jakarta` instead. We can use following dependencies to resolve the issue:
```
    swagger 'jakarta.servlet:jakarta.servlet-api:6.1.0'
    swagger 'io.swagger.core.v3:swagger-jaxrs2-jakarta:2.2.25'
```

**Comment by cshannon** [comment]
I am hoping the docs work to get the new Jakarta jar working, I don't think it's a good idea to go backwards and have to bring back the old javax dependency if it can be avoided. 

I don't think the javax approach even would work anyways, it would fix the error but it wouldn't find the right annotations when scanning the rest API as we migrated to Jakarta.

I can also try and find some time to look into it more later and create a PR if needed later this weekend or this upcoming week.

File: build.gradle (hunk 2 above)
Comment (cshannon) [reply build.gradle]: > @cshannon you are right - it is better to use `swagger-jaxrs2-jakarta` instead. We can use following dependencies to resolve the issue:
> 
> ```
> 
>     swagger 'jakarta.servlet:jakarta.servlet-api:6.1.0'
> 
>     swagger 'io.swagger.core.v3:swagger-jaxrs2-jakarta:2.2.25'
> 
> ```

That's good news, sorry for missing that on the original PR, I didn't realize it was used for the swagger generation.

Comment (m1a2st) [reply build.gradle]: Thank for @chia7712, @cshannon suggestion, addressed it

Comment (chia7712) [reply build.gradle]: > That's good news, sorry for missing that on the original PR, I didn't realize it was used for the swagger generation.

That's completely fine. Thank you for helping us resolve the issue! :)

**Comment by m1a2st** [comment]
I use `./gradlew clean releaseTarGz` in my local, it can build success
image

File: build.gradle [hunk 3]
```diff
@@ -3539,7 +3539,8 @@ project(':connect:runtime') {
 
     // We use this library to generate OpenAPI docs for the REST API, but we don't want or need it at compile
     // or run time. So, we add it to a separate configuration, which we use later on during docs generation
-    swagger libs.swaggerJaxrs2
```
Comment (chia7712) [inline build.gradle:3542]: `swaggerJaxrs2` is useless now, so could you please remove it ?

File: build.gradle [hunk 4]
```diff
@@ -3539,7 +3539,8 @@ project(':connect:runtime') {
...
+    swagger 'jakarta.servlet:jakarta.servlet-api:6.1.0'
+    swagger 'io.swagger.core.v3:swagger-jaxrs2-jakarta:2.2.25'
```
Comment (chia7712) [inline build.gradle]: we should add those to dependencies file

File: gradle/dependencies.gradle [hunk 5]
```diff
@@ -231,11 +233,12 @@ libs += [
...
   swaggerAnnotations: "io.swagger.core.v3:swagger-annotations:$swaggerVersion",
-  swaggerJaxrs2: "io.swagger.core.v3:swagger-jaxrs2:$swaggerVersion",
   zookeeper: "org.apache.zookeeper:zookeeper:$versions.zookeeper",
   jfreechart: "jfreechart:jfreechart:$versions.jfreechart",
   mavenArtifact: "org.apache.maven:maven-artifact:$versions.mavenArtifact",
   zstd: "com.github.luben:zstd-jni:$versions.zstd",
   httpclient: "org.apache.httpcomponents:httpclient:$versions.httpclient",
-  hdrHistogram: "org.hdrhistogram:HdrHistogram:$versions.hdrHistogram"
+  hdrHistogram: "org.hdrhistogram:HdrHistogram:$versions.hdrHistogram",
+  jakartaServletApi: "jakarta.servlet:jakarta.servlet-api:$versions.jakartaServletApi",
```
Comment (chia7712) [inline gradle/dependencies.gradle]: Could you please do a bit sort for those two new deps?

File: gradle/dependencies.gradle [hunk 6]
```diff
@@ -133,7 +133,9 @@ versions += [
   // Also make sure the compression levels in org.apache.kafka.common.record.CompressionType are still valid
   zstd: "1.5.6-6",
   junitPlatform: "1.10.2",
-  hdrHistogram: "2.2.2"
+  hdrHistogram: "2.2.2",
+  jakartaServletApi: "6.1.0",
```
Comment (chia7712) [inline gradle/dependencies.gradle]: ditto

**Comment by m1a2st** [comment]
I test in my local, I will generate `connect_rest.yaml` success
image

**Comment by mumrah** [comment]
@chia7712 how did we miss the breakage in the last PR? Do we need to add an explicit step to run `releaseTarGz`?

**Comment by chia7712** [comment]
> how did we miss the breakage in the last PR? Do we need to add an explicit step to run releaseTarGz?

The CI does not include building the documentation—maybe we should include it? for example, run `siteDocsTar`

**Comment by m1a2st** [comment]
If we need to add `siteDocsTar` on CI, I can help this issue :)

**Comment by chia7712** [comment]
> If we need to add siteDocsTar on CI, I can help this issue :)

I have opened [KAFKA-18241](https://issues.apache.org/jira/browse/KAFKA-18241), and it has been assigned to you. We can continue the discussion there.

**Comment by mumrah** [comment]
Yea let's do that as part of the "check -x test" step

**Review by chia7712 — APPROVED** [review]

File: gradle/dependencies.gradle [hunk 7]
```diff
@@ -166,6 +168,8 @@ libs += [
   jaxAnnotationApi: "javax.annotation:javax.annotation-api:$versions.jaxAnnotation",
   jaxbApi: "javax.xml.bind:jaxb-api:$versions.jaxb",
   jakartaRsApi: "jakarta.ws.rs:jakarta.ws.rs-api:$versions.jakartaRs",
+  jakartaServletApi: "jakarta.servlet:jakarta.servlet-api:$versions.jakartaServletApi",
+  jaxrs2Jakarta: "io.swagger.core.v3:swagger-jaxrs2-jakarta:$versions.jaxrs2Jakarta",
```
Comment (chia7712) [inline gradle/dependencies.gradle]: please use `swaggerVersion` for consistency

**Pull request merged**